### PR TITLE
Host header should contain host with port

### PR DESCRIPTION
**Pull request by obecker**
See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host

Required for https://github.com/http4k/http4k-connect/issues/361

### CODE REVIEW

File: http4k-core/src/main/kotlin/org/http4k/core/Uri.kt [hunk 1]
```diff
@@ -30,12 +30,16 @@ data class Uri(val scheme: String, val userInfo: String, val host: String, val p
...
-        .appendIfNotBlank(userInfo, userInfo, "@")
+    val hostAndPort = StringBuilder()
         .appendIfNotBlank(host, host)
         .appendIfPresent(port, ":", port.toString())
         .toString()
 
+    val authority = StringBuilder()
+        .appendIfNotBlank(userInfo, userInfo, "@")
+        .append(hostAndPort)
+        .toString()
```
Comment (obecker) [inline http4k-core/src/main/kotlin/org/http4k/core/Uri.kt:41]: I considered making them `lazy`, but I suppose it's not a big difference performance-wise.

**Comment by daviddenton** [comment]
We've had a chat and decided to keep the hostAndPort property out of the Uri class. This does involve recreating duplication, but it's targeted duplication so we're ok with it for now and we now feel happier to merge it in to unblock the problem in connect. Thanks for the contrib again! We'll get both out ASAP 😄

File: http4k-core/src/main/kotlin/org/http4k/filter/ClientFilters.kt [hunk 2]
```diff
@@ -182,7 +182,7 @@ object ClientFilters {
             val redirect = ensureValidMethodForRedirect().uri(newUri)
 
             return when {
-                header("host") != null && newUri.host.isNotEmpty() -> redirect.replaceHeader("host", newUri.hostAndPort)
+                header("host") != null && newUri.host.isNotEmpty() -> redirect.replaceHeader("host", newUri.host)
```
Comment (obecker) [inline http4k-core/src/main/kotlin/org/http4k/filter/ClientFilters.kt:185]: What about this one? Here a `host` header is set, so shouldn't it include the `port` as well?

**Comment by obecker** [comment]
You could create an extension function for `hostAndPort` to keep it out of the `Uri` class ...

File: http4k-core/src/main/kotlin/org/http4k/filter/ClientFilters.kt (hunk 2 above)
Comment (daviddenton) [reply http4k-core/src/main/kotlin/org/http4k/filter/ClientFilters.kt:185]: Missed that one. Will clean it up for the next release 🙃

**Pull request merged**